sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageStreamSource.java

```diff
@@ -210,20 +218,28 @@ private synchronized boolean readNextRecord() throws IOException {
           return false;
         }
 
-        // N.B.: For simplicity, we update fractionConsumed once a new response is fetched, not
-        // when we reach the end of the current response. In practice, this choice is not
-        // consequential.
-        fractionConsumed = fractionConsumedFromLastResponse;
+        fractionConsumedFromPreviousResponse = fractionConsumedFromCurrentResponse;
         ReadRowsResponse nextResponse = responseIterator.next();
         decoder =
             DecoderFactory.get()
                 .binaryDecoder(
                     nextResponse.getAvroRows().getSerializedBinaryRows().toByteArray(), decoder);
-        fractionConsumedFromLastResponse = getFractionConsumed(nextResponse);
+        rowsReadFromCurrentResponse = 0L;
+        rowCountFromCurrentResponse = nextResponse.getAvroRows().getRowCount();
+        fractionConsumedFromCurrentResponse = getFractionConsumed(nextResponse);
       }
 
       record = datumReader.read(record, decoder);
       current = parseFn.apply(new SchemaAndRecord(record, tableSchema));
+
+      rowsReadFromCurrentResponse++;
+      fractionConsumed =
+          fractionConsumedFromPreviousResponse
+              + (fractionConsumedFromCurrentResponse - fractionConsumedFromPreviousResponse)
+                  * rowsReadFromCurrentResponse
+                  * 1.0
+                  / rowCountFromCurrentResponse;
```

Comment:
Should this (rowCountFromCurrentResponse) be totalRowCountOfCurrentResponse ?

Reply:
That is a clearer name. Done.